File: server/src/utils/identity.ts
const RESERVED_IDENTIFIERS = new Set([
  "admin",
  "administrator",
  "owner",
  "root",
  "system",
  "support",
  "moder",
  "mod",
  "moderator",
  "administator",
]);

export const normalizeIdentity = (value: string) => value.trim().toLowerCase();

export const isReservedIdentity = (value: string) => RESERVED_IDENTIFIERS.has(normalizeIdentity(value));

export const getReservedIdentityError = (kind: "login" | "username") =>
  kind === "login" ? "Reserved login" : "Reserved username";

export const assertIdentityIsAllowed = (value: string, kind: "login" | "username") => {
  if (isReservedIdentity(value)) {
    throw new Error(getReservedIdentityError(kind));
  }
};

File: server-go/utils/identity.go
package utils

import "strings"

var reservedIdentifiers = map[string]bool{
	"admin":         true,
	"administrator": true,
	"owner":         true,
	"root":          true,
	"system":        true,
	"support":       true,
	"moder":         true,
	"mod":           true,
	"moderator":     true,
	"administator":  true, // protect against anti-phishing misspelling
}

// NormalizeIdentity trims and lowercases the input string
func NormalizeIdentity(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}

// IsReservedIdentity returns true if the input identity is reserved
func IsReservedIdentity(value string) bool {
	return reservedIdentifiers[NormalizeIdentity(value)]
}
